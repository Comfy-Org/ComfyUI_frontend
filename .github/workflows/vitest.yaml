name: Vitest Tests

on:
  push:
    branches: [ main, master, dev* ]
  pull_request:
    branches: [ main, master, dev* ]

jobs:
  test:
    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v4

    - name: Use Node.js
      uses: actions/setup-node@v4
      with:
        node-version: 'lts/*'

    - name: Install dependencies
      run: npm ci

    - name: Run Vitest tests
      run: npm run test:component
