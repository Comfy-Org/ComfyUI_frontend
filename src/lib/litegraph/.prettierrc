{
  "singleQuote": false,
  "semi": false,
  "tabWidth": 2,
  "trailingComma": "all",
  "overrides": [{ "files": "*.ts", "options": { "requirePragma": true } }],
  "arrowParens": "avoid"
}
