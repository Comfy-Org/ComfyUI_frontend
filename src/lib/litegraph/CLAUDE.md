- This codebase has extensive eslint autofix rules and IDEs are configured to use eslint as the format on save tool. Run ESLint instead of manually figuring out whitespace fixes or other trivial style concerns. Review the results and correct any remaining eslint errors.
- Take advantage of `TypedArray` `subarray` when appropriate.
- The `size` and `pos` properties of `Rectangle` share the same array buffer (`subarray`); they may be used to set the rectangles size and position.
- Prefer single line `if` syntax over adding curly braces, when the statement has a very concise expression and concise, single line statement.
- Do not replace `&&=` or `||=` with `=` when there is no reason to do so. If you do find a reason to remove either `&&=` or `||=`, leave a comment explaining why the removal occurred.
- You are allowed to research code on https://developer.mozilla.org/ and https://stackoverflow.com without asking.
- When adding features, always write vitest unit tests using cursor rules in @.cursor
- When writing methods, prefer returning idiomatic JavaScript `undefined` over `null`.

# Bash commands

- `npm run typecheck` Run the typechecker
- `npm run build` Build the project
- `npm run lint:fix` Run ESLint

# Code style

- Always prefer best practices when writing code.
- Write using concise, legible, and easily maintainable code.
- Avoid repetition where possible, but not at the expense of code legibility.
- Type assertions are an absolute last resort. In almost all cases, they are a crutch that leads to brittle code.

# Workflow

- Be sure to typecheck when youâ€™re done making a series of code changes
- Prefer running single tests, and not the whole test suite, for performance
