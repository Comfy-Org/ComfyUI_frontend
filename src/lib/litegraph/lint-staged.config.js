export default {
  "*.css": stagedFiles => `prettier --write ${stagedFiles.join(" ")}`,

  "*.js": stagedFiles => prettierAndEslint(stagedFiles),

  "*.{ts,mts}": stagedFiles => [
    ...prettierAndEslint(stagedFiles),
    "tsc",
  ],
}

function prettierAndEslint(fileNames) {
  return [
    `prettier --write ${fileNames.join(" ")}`,
    `eslint --fix ${fileNames.join(" ")}`,
  ]
}
