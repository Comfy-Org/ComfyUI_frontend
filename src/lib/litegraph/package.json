{
  "name": "@comfyorg/litegraph",
  "version": "0.16.3",
  "type": "module",
  "description": "A graph node editor similar to PD or UDK Blueprints. It works in an HTML5 Canvas and allows to export graphs to be included in applications.",
  "main": "./dist/litegraph.umd.js",
  "module": "./dist/litegraph.es.js",
  "types": "./dist/litegraph.d.ts",
  "style": "./dist/css/litegraph.css",
  "exports": {
    ".": {
      "types": "./dist/litegraph.d.ts",
      "import": "./dist/litegraph.es.js",
      "require": "./dist/litegraph.umd.js"
    },
    "./style.css": "./dist/css/litegraph.css",
    "./dist/*": "./dist/*.d.ts"
  },
  "directories": {
    "doc": "doc"
  },
  "private": false,
  "scripts": {
    "build": "tsc && vite build",
    "dev": "vite",
    "preview": "vite preview",
    "watch": "vite build --watch",
    "test": "vitest",
    "lint": "eslint",
    "lint:fix": "eslint --fix",
    "lint:ci": "eslint src",
    "format": "prettier --check './src/*.{js,ts,tsx,vue,mts}'",
    "format:fix": "prettier --write './src/*.{js,ts,tsx,vue,mts}'",
    "typecheck": "tsc"
  },
  "repository": {
    "type": "git",
    "url": "git@github.com:Comfy-Org/litegraph.js.git"
  },
  "author": "comfyorg",
  "license": "MIT",
  "files": [
    "dist"
  ],
  "bugs": {
    "url": "https://github.com/Comfy-Org/litegraph.js/issues"
  },
  "homepage": "https://github.com/Comfy-Org/litegraph.js",
  "devDependencies": {
    "@types/node": "^22.13.9",
    "jsdom": "^25.0.1",
    "ts-node": "^10.9.2",
    "vite-plugin-dts": "^4.5.4",
  }
}
