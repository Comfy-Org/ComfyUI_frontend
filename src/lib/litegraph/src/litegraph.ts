import type { ContextMenu } from "./ContextMenu"
import type { ConnectingLink, Point } from "./interfaces"
import type {
  IContextMenuOptions,
  INodeSlot,
  Size,
} from "./interfaces"
import type { LGraphNode } from "./LGraphNode"
import type { CanvasEventDetail } from "./types/events"
import type { RenderShape, TitleMode } from "./types/globalEnums"

// Must remain above LiteGraphGlobal (circular dependency due to abstract factory behaviour in `configure`)
export { Subgraph } from "./subgraph/Subgraph"

import { LiteGraphGlobal } from "./LiteGraphGlobal"
import { loadPolyfills } from "./polyfills"

export const LiteGraph = new LiteGraphGlobal()

// Load legacy polyfills
loadPolyfills()

// Backwards compat

// Type definitions for litegraph.js 0.7.0
// Project: litegraph.js
// Definitions by: NateScarlet <https://github.com/NateScarlet>
/** @deprecated Use {@link Point} instead. */
export type Vector2 = Point
/** @deprecated Use {@link Rect} instead. */
export type Vector4 = [number, number, number, number]

export interface IContextMenuItem {
  content: string
  callback?: ContextMenuEventListener
  /** Used as innerHTML for extra child element */
  title?: string
  disabled?: boolean
  has_submenu?: boolean
  submenu?: {
    options: IContextMenuItem[]
  } & IContextMenuOptions
  className?: string
}

export type ContextMenuEventListener = (
  value: IContextMenuItem,
  options: IContextMenuOptions,
  event: MouseEvent,
  parentMenu: ContextMenu<unknown> | undefined,
  node: LGraphNode,
) => boolean | void

export interface LinkReleaseContext {
  node_to?: LGraphNode
  node_from?: LGraphNode
  slot_from: INodeSlot
  type_filter_in?: string
  type_filter_out?: string
}

export interface LinkReleaseContextExtended {
  links: ConnectingLink[]
}

export interface LiteGraphCanvasEvent extends CustomEvent<CanvasEventDetail> {}

export interface LGraphNodeConstructor<T extends LGraphNode = LGraphNode> {
  new (title: string, type?: string): T

  title: string
  type: string
  size?: Size
  min_height?: number
  slot_start_y?: number
  widgets_info?: any
  collapsable?: boolean
  color?: string
  bgcolor?: string
  shape?: RenderShape
  title_mode?: TitleMode
  title_color?: string
  title_text_color?: string
  keepAllLinksOnBypass: boolean
  resizeHandleSize?: number
  resizeEdgeSize?: number
}

// End backwards compat

export { InputIndicators } from "./canvas/InputIndicators"
export { LinkConnector } from "./canvas/LinkConnector"
export { isOverNodeInput, isOverNodeOutput } from "./canvas/measureSlots"
export { CanvasPointer } from "./CanvasPointer"
export * as Constants from "./constants"
export { ContextMenu } from "./ContextMenu"
export { CurveEditor } from "./CurveEditor"
export { DragAndScale } from "./DragAndScale"
export { LabelPosition, SlotDirection, SlotShape, SlotType } from "./draw"
export { strokeShape } from "./draw"
export { Rectangle } from "./infrastructure/Rectangle"
export type {
  CanvasColour,
  ColorOption,
  ConnectingLink,
  Direction,
  IBoundaryNodes,
  IColorable,
  IContextMenuOptions,
  IContextMenuValue,
  IFoundSlot,
  IInputOrOutput,
  INodeFlags,
  INodeInputSlot,
  INodeOutputSlot,
  INodeSlot,
  ISlotType,
  KeysOfType,
  LinkNetwork,
  LinkSegment,
  MethodNames,
  PickByType,
  Point,
  Positionable,
  ReadonlyLinkNetwork,
  ReadOnlyPoint,
  ReadOnlyRect,
  Rect,
  Size,
} from "./interfaces"
export { LGraph } from "./LGraph"
export { BadgePosition, LGraphBadge, type LGraphBadgeOptions } from "./LGraphBadge"
export { LGraphCanvas, type LGraphCanvasState } from "./LGraphCanvas"
export { LGraphGroup } from "./LGraphGroup"
export { LGraphNode, type NodeId } from "./LGraphNode"
export { COMFY_VUE_NODE_DIMENSIONS, type ComponentHeightKey } from "./LiteGraphGlobal"
export { LGraphNodeProperties } from "./LGraphNodeProperties"
export { type LinkId, LLink } from "./LLink"
export { clamp, createBounds } from "./measure"
export { Reroute, type RerouteId } from "./Reroute"
export { type ExecutableLGraphNode, ExecutableNodeDTO } from "./subgraph/ExecutableNodeDTO"
export { SubgraphNode } from "./subgraph/SubgraphNode"
export type { CanvasPointerEvent } from "./types/events"
export {
  CanvasItem,
  EaseFunction,
  LGraphEventMode,
  LinkMarkerShape,
  RenderShape,
  TitleMode,
} from "./types/globalEnums"
export type {
  ExportedSubgraph,
  ExportedSubgraphInstance,
  ExportedSubgraphIONode,
  ISerialisedGraph,
  SerialisableGraph,
  SerialisableLLink,
  SubgraphIO,
} from "./types/serialisation"
export type { IWidget } from "./types/widgets"
export { isColorable } from "./utils/type"
export { createUuidv4 } from "./utils/uuid"
export { BaseSteppedWidget } from "./widgets/BaseSteppedWidget"
export { BaseWidget } from "./widgets/BaseWidget"
export { BooleanWidget } from "./widgets/BooleanWidget"
export { ButtonWidget } from "./widgets/ButtonWidget"
export { ComboWidget } from "./widgets/ComboWidget"
export { KnobWidget } from "./widgets/KnobWidget"
export { LegacyWidget } from "./widgets/LegacyWidget"
export { NumberWidget } from "./widgets/NumberWidget"
export { SliderWidget } from "./widgets/SliderWidget"
export { TextWidget } from "./widgets/TextWidget"
export { isComboWidget } from "./widgets/widgetMap"
