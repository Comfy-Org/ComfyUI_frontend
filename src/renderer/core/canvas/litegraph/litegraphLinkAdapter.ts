/**
 * Litegraph Link Adapter
 *
 * Bridges the gap between litegraph's data model and the pure canvas renderer.
 * Converts litegraph-specific types (LLink, LGraphNode, slots) into generic
 * rendering data that can be consumed by the PathRenderer.
 * Maintains backward compatibility with existing litegraph integration.
 */
import type { LLink } from '@/lib/litegraph/src/LLink'
import type { Reroute } from '@/lib/litegraph/src/Reroute'
import type {
  CanvasColour,
  ReadOnlyPoint
} from '@/lib/litegraph/src/interfaces'
import { LiteGraph } from '@/lib/litegraph/src/litegraph'
import {
  LinkDirection,
  LinkMarkerShape,
  LinkRenderType
} from '@/lib/litegraph/src/types/globalEnums'
import {
  type ArrowShape,
  CanvasPathRenderer,
  type Direction,
  type LinkRenderData,
  type RenderContext as PathRenderContext,
  type Point,
  type RenderMode
} from '@/renderer/core/canvas/pathRenderer'
import { layoutStore } from '@/renderer/core/layout/store/layoutStore'
import type { Bounds } from '@/renderer/core/layout/types'

export interface LinkRenderContext {
  // Canvas settings
  renderMode: LinkRenderType
  connectionWidth: number
  renderBorder: boolean
  lowQuality: boolean
  highQualityRender: boolean
  scale: number
  linkMarkerShape: LinkMarkerShape
  renderConnectionArrows: boolean

  // State
  highlightedLinks: Set<string | number>

  // Colors
  defaultLinkColor: CanvasColour
  linkTypeColors: Record<string, CanvasColour>

  // Pattern for disabled links (optional)
  disabledPattern?: CanvasPattern | null
}

export class LitegraphLinkAdapter {
  private readonly pathRenderer = new CanvasPathRenderer()

  constructor(public readonly enableLayoutStoreWrites = true) {}

  /**
   * Convert LinkDirection enum to Direction string
   */
  private convertDirection(dir: LinkDirection): Direction {
    switch (dir) {
      case LinkDirection.LEFT:
        return 'left'
      case LinkDirection.RIGHT:
        return 'right'
      case LinkDirection.UP:
        return 'up'
      case LinkDirection.DOWN:
        return 'down'
      case LinkDirection.CENTER:
        return 'none'
      default:
        return 'right'
    }
  }

  /**
   * Convert LinkRenderContext to PathRenderContext
   */
  private convertToPathRenderContext(
    context: LinkRenderContext
  ): PathRenderContext {
    // Match original arrow rendering conditions:
    // Arrows only render when scale >= 0.6 AND highquality_render AND render_connection_arrows
    const shouldShowArrows =
      context.scale >= 0.6 &&
      context.highQualityRender &&
      context.renderConnectionArrows

    // Only show center marker when not set to None
    const shouldShowCenterMarker =
      context.linkMarkerShape !== LinkMarkerShape.None

    return {
      style: {
        mode: this.convertRenderMode(context.renderMode),
        connectionWidth: context.connectionWidth,
        borderWidth: context.renderBorder ? 4 : undefined,
        arrowShape: this.convertArrowShape(context.linkMarkerShape),
        showArrows: shouldShowArrows,
        lowQuality: context.lowQuality,
        // Center marker settings (matches original litegraph behavior)
        showCenterMarker: shouldShowCenterMarker,
        centerMarkerShape:
          context.linkMarkerShape === LinkMarkerShape.Arrow
            ? 'arrow'
            : 'circle',
        highQuality: context.highQualityRender
      },
      colors: {
        default: String(context.defaultLinkColor),
        byType: this.convertColorMap(context.linkTypeColors),
        highlighted: '#FFF'
      },
      patterns: {
        disabled: context.disabledPattern
      },
      animation: {
        time: LiteGraph.getTime() * 0.001
      },
      scale: context.scale,
      highlightedIds: new Set(Array.from(context.highlightedLinks).map(String))
    }
  }

  /**
   * Convert LinkRenderType to RenderMode
   */
  private convertRenderMode(mode: LinkRenderType): RenderMode {
    switch (mode) {
      case LinkRenderType.LINEAR_LINK:
        return 'linear'
      case LinkRenderType.STRAIGHT_LINK:
        return 'straight'
      case LinkRenderType.SPLINE_LINK:
      default:
        return 'spline'
    }
  }

  /**
   * Convert LinkMarkerShape to ArrowShape
   */
  private convertArrowShape(shape: LinkMarkerShape): ArrowShape {
    switch (shape) {
      case LinkMarkerShape.Circle:
        return 'circle'
      case LinkMarkerShape.Arrow:
      default:
        return 'triangle'
    }
  }

  /**
   * Convert color map to ensure all values are strings
   */
  private convertColorMap(
    colors: Record<string, CanvasColour>
  ): Record<string, string> {
    const result: Record<string, string> = {}
    for (const [key, value] of Object.entries(colors)) {
      result[key] = String(value)
    }
    return result
  }

  /**
   * Apply spline offset to a point, mimicking original #addSplineOffset behavior
   * Critically: does nothing for CENTER/NONE directions (no case for them)
   */
  private applySplineOffset(
    point: Point,
    direction: LinkDirection,
    distance: number
  ): void {
    switch (direction) {
      case LinkDirection.LEFT:
        point.x -= distance
        break
      case LinkDirection.RIGHT:
        point.x += distance
        break
      case LinkDirection.UP:
        point.y -= distance
        break
      case LinkDirection.DOWN:
        point.y += distance
        break
      // CENTER and NONE: no offset applied (original behavior)
    }
  }

  /**
   * Direct rendering method compatible with LGraphCanvas
   * Converts data and delegates to pure renderer
   */
  renderLinkDirect(
    ctx: CanvasRenderingContext2D,
    a: ReadOnlyPoint,
    b: ReadOnlyPoint,
    link: LLink | null,
    skip_border: boolean,
    flow: number | boolean | null,
    color: CanvasColour | null,
    start_dir: LinkDirection,
    end_dir: LinkDirection,
    context: LinkRenderContext,
    extras: {
      reroute?: Reroute
      startControl?: ReadOnlyPoint
      endControl?: ReadOnlyPoint
      num_sublines?: number
      disabled?: boolean
    } = {}
  ): void {
    // Apply same defaults as original renderLink
    const startDir = start_dir || LinkDirection.RIGHT
    const endDir = end_dir || LinkDirection.LEFT

    // Convert flow to boolean
    const flowBool = flow === true || (typeof flow === 'number' && flow > 0)

    // Create LinkRenderData from direct parameters
    const linkData: LinkRenderData = {
      id: link ? String(link.id) : 'temp',
      startPoint: { x: a[0], y: a[1] },
      endPoint: { x: b[0], y: b[1] },
      startDirection: this.convertDirection(startDir),
      endDirection: this.convertDirection(endDir),
      color: color !== null && color !== undefined ? String(color) : undefined,
      type: link?.type !== undefined ? String(link.type) : undefined,
      flow: flowBool,
      disabled: extras.disabled || false
    }

    // Control points handling (spline mode):
    // - Pre-refactor, the old renderLink honored a single provided control and
    //   derived the missing side via #addSplineOffset (CENTER => no offset).
    // - Restore that behavior here so reroute segments render identically.
    if (context.renderMode === LinkRenderType.SPLINE_LINK) {
      const hasStartCtrl = !!extras.startControl
      const hasEndCtrl = !!extras.endControl

      // Compute distance once for offsets
      const dist = Math.sqrt(
        (b[0] - a[0]) * (b[0] - a[0]) + (b[1] - a[1]) * (b[1] - a[1])
      )
      const factor = 0.25

      const cps: Point[] = []

      if (hasStartCtrl && hasEndCtrl) {
        // Both provided explicitly
        cps.push(
          {
            x: a[0] + (extras.startControl![0] || 0),
            y: a[1] + (extras.startControl![1] || 0)
          },
          {
            x: b[0] + (extras.endControl![0] || 0),
            y: b[1] + (extras.endControl![1] || 0)
          }
        )
        linkData.controlPoints = cps
      } else if (hasStartCtrl && !hasEndCtrl) {
        // Start provided, derive end via direction offset (CENTER => no offset)
        const start = {
          x: a[0] + (extras.startControl![0] || 0),
          y: a[1] + (extras.startControl![1] || 0)
        }
        const end = { x: b[0], y: b[1] }
        this.applySplineOffset(end, endDir, dist * factor)
        cps.push(start, end)
        linkData.controlPoints = cps
      } else if (!hasStartCtrl && hasEndCtrl) {
        // End provided, derive start via direction offset (CENTER => no offset)
        const start = { x: a[0], y: a[1] }
        this.applySplineOffset(start, startDir, dist * factor)
        const end = {
          x: b[0] + (extras.endControl![0] || 0),
          y: b[1] + (extras.endControl![1] || 0)
        }
        cps.push(start, end)
        linkData.controlPoints = cps
      } else {
        // Neither provided: derive both from directions (CENTER => no offset)
        const start = { x: a[0], y: a[1] }
        const end = { x: b[0], y: b[1] }
        this.applySplineOffset(start, startDir, dist * factor)
        this.applySplineOffset(end, endDir, dist * factor)
        cps.push(start, end)
        linkData.controlPoints = cps
      }
    }

    // Convert context
    const pathContext = this.convertToPathRenderContext(context)

    // Override skip_border if needed
    if (skip_border) {
      pathContext.style.borderWidth = undefined
    }

    // Render using pure renderer
    const path = this.pathRenderer.drawLink(ctx, linkData, pathContext)

    // Store path for hit detection
    const linkSegment = extras.reroute ?? link
    if (linkSegment) {
      linkSegment.path = path

      // Copy calculated center position back to litegraph object
      // This is needed for hit detection and menu interaction
      if (linkData.centerPos) {
        linkSegment._pos = linkSegment._pos || new Float32Array(2)
        linkSegment._pos[0] = linkData.centerPos.x
        linkSegment._pos[1] = linkData.centerPos.y

        // Store center angle if calculated (for arrow markers)
        if (linkData.centerAngle !== undefined) {
          linkSegment._centreAngle = linkData.centerAngle
        }
      }

      // Update layout store when writes are enabled (event-driven path)
      if (this.enableLayoutStoreWrites && link && link.id !== -1) {
        // Calculate bounds and center only when writing
        const bounds = this.calculateLinkBounds(
          [linkData.startPoint.x, linkData.startPoint.y] as ReadOnlyPoint,
          [linkData.endPoint.x, linkData.endPoint.y] as ReadOnlyPoint,
          linkData
        )
        const centerPos = linkData.centerPos || {
          x: (linkData.startPoint.x + linkData.endPoint.x) / 2,
          y: (linkData.startPoint.y + linkData.endPoint.y) / 2
        }

        // Update whole link layout (only if not a reroute segment)
        if (!extras.reroute) {
          layoutStore.updateLinkLayout(link.id, {
            id: link.id,
            path: path,
            bounds: bounds,
            centerPos: centerPos,
            sourceNodeId: String(link.origin_id),
            targetNodeId: String(link.target_id),
            sourceSlot: link.origin_slot,
            targetSlot: link.target_slot
          })
        }

        // Always update segment layout (for both regular links and reroute segments)
        const rerouteId = extras.reroute ? extras.reroute.id : null
        layoutStore.updateLinkSegmentLayout(link.id, rerouteId, {
          path: path,
          bounds: bounds,
          centerPos: centerPos
        })
      }
    }
  }

  renderDraggingLink(
    ctx: CanvasRenderingContext2D,
    from: ReadOnlyPoint,
    to: ReadOnlyPoint,
    colour: CanvasColour,
    startDir: LinkDirection,
    endDir: LinkDirection,
    context: LinkRenderContext
  ): void {
    this.renderLinkDirect(
      ctx,
      from,
      to,
      null,
      false,
      null,
      colour,
      startDir,
      endDir,
      {
        ...context,
        linkMarkerShape: LinkMarkerShape.None
      },
      {
        disabled: false
      }
    )
  }

  /**
   * Calculate bounding box for a link
   * Includes padding for line width and control points
   */
  private calculateLinkBounds(
    startPos: ReadOnlyPoint,
    endPos: ReadOnlyPoint,
    linkData: LinkRenderData
  ): Bounds {
    let minX = Math.min(startPos[0], endPos[0])
    let maxX = Math.max(startPos[0], endPos[0])
    let minY = Math.min(startPos[1], endPos[1])
    let maxY = Math.max(startPos[1], endPos[1])

    // Include control points if they exist (for spline links)
    if (linkData.controlPoints) {
      for (const cp of linkData.controlPoints) {
        minX = Math.min(minX, cp.x)
        maxX = Math.max(maxX, cp.x)
        minY = Math.min(minY, cp.y)
        maxY = Math.max(maxY, cp.y)
      }
    }

    // Add padding for line width and hit tolerance
    const padding = 20

    return {
      x: minX - padding,
      y: minY - padding,
      width: maxX - minX + 2 * padding,
      height: maxY - minY + 2 * padding
    }
  }
}
