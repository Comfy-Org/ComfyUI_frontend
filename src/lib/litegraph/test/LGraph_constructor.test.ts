import { describe } from "vitest"

import { LGraph } from "@/litegraph"

import { dirtyTest } from "./testExtensions"

describe("LGraph (constructor only)", () => {
  dirtyTest("Matches previous snapshot", ({ expect, minimalSerialisableGraph, basicSerialisableGraph }) => {
    const minLGraph = new LGraph(minimalSerialisableGraph)
    expect(minLGraph).toMatchSnapshot("minLGraph")

    const basicLGraph = new LGraph(basicSerialisableGraph)
    expect(basicLGraph).toMatchSnapshot("basicLGraph")
  })
})
