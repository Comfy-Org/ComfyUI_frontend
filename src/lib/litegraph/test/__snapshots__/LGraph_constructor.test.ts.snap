// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LGraph (constructor only) > Matches previous snapshot > basicLGraph 1`] = `
LGraph {
  "_groups": [
    LGraphGroup {
      "_bounding": Float32Array [
        20,
        20,
        1,
        3,
      ],
      "_children": Set {},
      "_nodes": [],
      "_pos": Float32Array [
        20,
        20,
      ],
      "_size": Float32Array [
        1,
        3,
      ],
      "color": "#6029aa",
      "flags": {},
      "font": undefined,
      "font_size": 14,
      "graph": [Circular],
      "id": 123,
      "isPointInside": [Function],
      "selected": undefined,
      "setDirtyCanvas": [Function],
      "title": "A group to test with",
    },
  ],
  "_input_nodes": undefined,
  "_last_trigger_time": undefined,
  "_links": Map {},
  "_nodes": [
    LGraphNode {
      "_collapsed_width": undefined,
      "_level": undefined,
      "_pos": Float32Array [
        10,
        10,
      ],
      "_posSize": Float32Array [
        10,
        10,
        140,
        60,
      ],
      "_relative_id": undefined,
      "_shape": undefined,
      "_size": Float32Array [
        140,
        60,
      ],
      "action_call": undefined,
      "action_triggered": undefined,
      "badgePosition": "top-left",
      "badges": [],
      "bgcolor": undefined,
      "block_delete": undefined,
      "boxcolor": undefined,
      "changeTracker": undefined,
      "clip_area": undefined,
      "clonable": undefined,
      "color": undefined,
      "console": undefined,
      "exec_version": undefined,
      "execute_triggered": undefined,
      "flags": {},
      "freeWidgetSpace": undefined,
      "gotFocusAt": undefined,
      "graph": [Circular],
      "has_errors": undefined,
      "id": 1,
      "ignore_remove": undefined,
      "inputs": [],
      "last_serialization": undefined,
      "locked": undefined,
      "lostFocusAt": undefined,
      "mode": 0,
      "mouseOver": undefined,
      "order": 0,
      "outputs": [],
      "progress": undefined,
      "properties": {},
      "properties_info": [],
      "redraw_on_mouse": undefined,
      "removable": undefined,
      "resizable": undefined,
      "selected": undefined,
      "serialize_widgets": undefined,
      "showAdvanced": undefined,
      "strokeStyles": {
        "error": [Function],
        "selected": [Function],
      },
      "title": "LGraphNode",
      "type": "mustBeSet",
      "widgets": undefined,
      "widgets_start_y": undefined,
      "widgets_up": undefined,
    },
  ],
  "_nodes_by_id": {
    "1": LGraphNode {
      "_collapsed_width": undefined,
      "_level": undefined,
      "_pos": Float32Array [
        10,
        10,
      ],
      "_posSize": Float32Array [
        10,
        10,
        140,
        60,
      ],
      "_relative_id": undefined,
      "_shape": undefined,
      "_size": Float32Array [
        140,
        60,
      ],
      "action_call": undefined,
      "action_triggered": undefined,
      "badgePosition": "top-left",
      "badges": [],
      "bgcolor": undefined,
      "block_delete": undefined,
      "boxcolor": undefined,
      "changeTracker": undefined,
      "clip_area": undefined,
      "clonable": undefined,
      "color": undefined,
      "console": undefined,
      "exec_version": undefined,
      "execute_triggered": undefined,
      "flags": {},
      "freeWidgetSpace": undefined,
      "gotFocusAt": undefined,
      "graph": [Circular],
      "has_errors": undefined,
      "id": 1,
      "ignore_remove": undefined,
      "inputs": [],
      "last_serialization": undefined,
      "locked": undefined,
      "lostFocusAt": undefined,
      "mode": 0,
      "mouseOver": undefined,
      "order": 0,
      "outputs": [],
      "progress": undefined,
      "properties": {},
      "properties_info": [],
      "redraw_on_mouse": undefined,
      "removable": undefined,
      "resizable": undefined,
      "selected": undefined,
      "serialize_widgets": undefined,
      "showAdvanced": undefined,
      "strokeStyles": {
        "error": [Function],
        "selected": [Function],
      },
      "title": "LGraphNode",
      "type": "mustBeSet",
      "widgets": undefined,
      "widgets_start_y": undefined,
      "widgets_up": undefined,
    },
  },
  "_nodes_executable": [],
  "_nodes_in_order": [
    LGraphNode {
      "_collapsed_width": undefined,
      "_level": undefined,
      "_pos": Float32Array [
        10,
        10,
      ],
      "_posSize": Float32Array [
        10,
        10,
        140,
        60,
      ],
      "_relative_id": undefined,
      "_shape": undefined,
      "_size": Float32Array [
        140,
        60,
      ],
      "action_call": undefined,
      "action_triggered": undefined,
      "badgePosition": "top-left",
      "badges": [],
      "bgcolor": undefined,
      "block_delete": undefined,
      "boxcolor": undefined,
      "changeTracker": undefined,
      "clip_area": undefined,
      "clonable": undefined,
      "color": undefined,
      "console": undefined,
      "exec_version": undefined,
      "execute_triggered": undefined,
      "flags": {},
      "freeWidgetSpace": undefined,
      "gotFocusAt": undefined,
      "graph": [Circular],
      "has_errors": undefined,
      "id": 1,
      "ignore_remove": undefined,
      "inputs": [],
      "last_serialization": undefined,
      "locked": undefined,
      "lostFocusAt": undefined,
      "mode": 0,
      "mouseOver": undefined,
      "order": 0,
      "outputs": [],
      "progress": undefined,
      "properties": {},
      "properties_info": [],
      "redraw_on_mouse": undefined,
      "removable": undefined,
      "resizable": undefined,
      "selected": undefined,
      "serialize_widgets": undefined,
      "showAdvanced": undefined,
      "strokeStyles": {
        "error": [Function],
        "selected": [Function],
      },
      "title": "LGraphNode",
      "type": "mustBeSet",
      "widgets": undefined,
      "widgets_start_y": undefined,
      "widgets_up": undefined,
    },
  ],
  "_subgraphs": Map {},
  "_version": 3,
  "catch_errors": true,
  "config": {},
  "elapsed_time": 0.01,
  "errors_in_execution": undefined,
  "events": CustomEventTarget {},
  "execution_time": undefined,
  "execution_timer_id": undefined,
  "extra": {},
  "filter": undefined,
  "fixedtime": 0,
  "fixedtime_lapse": 0.01,
  "globaltime": 0,
  "id": "ca9da7d8-fddd-4707-ad32-67be9be13140",
  "iteration": 0,
  "last_update_time": 0,
  "links": Map {},
  "list_of_graphcanvas": null,
  "nodes_actioning": [],
  "nodes_executedAction": [],
  "nodes_executing": [],
  "revision": 0,
  "runningtime": 0,
  "starttime": 0,
  "state": {
    "lastGroupId": 123,
    "lastLinkId": 0,
    "lastNodeId": 1,
    "lastRerouteId": 0,
  },
  "status": 1,
  "vars": {},
  "version": 1,
}
`;

exports[`LGraph (constructor only) > Matches previous snapshot > minLGraph 1`] = `
LGraph {
  "_groups": [],
  "_input_nodes": undefined,
  "_last_trigger_time": undefined,
  "_links": Map {},
  "_nodes": [],
  "_nodes_by_id": {},
  "_nodes_executable": [],
  "_nodes_in_order": [],
  "_subgraphs": Map {},
  "_version": 0,
  "catch_errors": true,
  "config": {},
  "elapsed_time": 0.01,
  "errors_in_execution": undefined,
  "events": CustomEventTarget {},
  "execution_time": undefined,
  "execution_timer_id": undefined,
  "extra": {},
  "filter": undefined,
  "fixedtime": 0,
  "fixedtime_lapse": 0.01,
  "globaltime": 0,
  "id": "d175890f-716a-4ece-ba33-1d17a513b7be",
  "iteration": 0,
  "last_update_time": 0,
  "links": Map {},
  "list_of_graphcanvas": null,
  "nodes_actioning": [],
  "nodes_executedAction": [],
  "nodes_executing": [],
  "revision": 0,
  "runningtime": 0,
  "starttime": 0,
  "state": {
    "lastGroupId": 0,
    "lastLinkId": 0,
    "lastNodeId": 0,
    "lastRerouteId": 0,
  },
  "status": 1,
  "vars": {},
  "version": 1,
}
`;
